Find builders by entry point

brechtm opened the pull request:
A builder is uniquely identified by its name, which can be used as an
entry point in the sphinx.builders entry point group. This obviates
registering the builder as an extension.

The built-in builders are now also located by means of entry points.
New third-party builders should provide an entry point (see setup.py
for examples), and define a setup(cls, app) class method to define their
configuration variables etc.

Existing builders can still be loaded the traditional way, by including
their module name in the extensions list in conf.py.

brechtm committed:
Find third-party builders by entry point

A builder is uniquely identified by its name, which can be used as an
entry point in the 'sphinx.builders' entry point group. This obviates
the need to register the builder as an extension.

The built-in builders are still loaded as before. New third-party builders
should provide an entry point in their setup.py:

    entry_points={
        'sphinx.builders': [
            'mybuilder = mypackage.mymodule:MyBuilder',
        ],
    }

Like before, builders should define a setup(app) function in the
'mypackage.module' module to define configuration variables etc. It is
no longer necessary to register the builder using Sphinx.add_builder().

Existing builders can still be loaded the traditional way, by including
their module name in the extensions list in conf.py.

tk0miya commented:
I'm interested in this. But this PR is conflicted. Could you fix this please?

BTW, is there any reason to support only builders? It would be nice if it supports all types of sphinx extensions.

brechtm commented:
I've rebased this branch onto the current master.

This only supports builders because they are uniquely defined by their name (the string passed to `sphinx-build`'s `-b` option. This is not the case for other types of extensions. You could assign a name to each extension, but I don't see the advantage over specifying extensions by name instead of by module path.

For builders, this is convenient because the builder no longer needs to be included in the list of extensions in `conf.py`. I believe the builder used is not an intrinsic part of the project's configuration unlike other extensions. It should be possible to clone a Sphinx project and build it using an arbitrary builder without requiring a change to its `conf.py`.

brechtm commented:
Are there any objections to merging this in?
I hope my discussion of why this is limited to builders was clear?

brechtm commented:
Can you have another look at this PR, please? What are the objections, if any? Thanks!

tk0miya commented on setup.py:
```diff
@@ -211,6 +211,31 @@ def _run_domain_js(self, domain):
+            'qthelp = sphinx.builders.qthelp:QtHelpBuilder',
+            'applehelp = sphinx.builders.applehelp:AppleHelpBuilder',
+            'epub = sphinx.builders.epub:EpubBuilder',
+            'epub3 = sphinx.builders.epub3:Epub3Builder',
```
Now we use Epub3builder as "epub" builder. So please rename EpubBuilder to epub2, and Epub3Builder to just epub.

brechtm replied:
Entry points for the built-in builders were removed as requested by @shimizukawa. This could be added back at a later point.

tk0miya commented on sphinx/application.py:
```diff
@@ -298,10 +299,18 @@ def _init_builder(self, buildername):
+            builderclass = self.builderclasses[buildername]
+        else:
+            builderclass = entry_point.load()
+            builderclass.setup(self)
```
Why don't you use `extenion_module.setup()`? A builder class is only a part of extensions. The builder might use event handlers, config values and so on. For example, LaTeX builder uses an event handler and some config values.
So I prefer to use the setup function of the extension module.

But, unfortunately, `_init_builder()` is called after initialization of config values.
So it is too late to load an extension. It would be nice to load before the initialization of config, and create an instance of builder after the initialization.

tk0miya replied:
BTW builtin builders are already loaded. They might be loaded twice.
It would be better to load them only if needed. So I'd like to remove them from list of builtin_extensions.

brechtm replied:
I'm now using ``extenion_module.setup()``. The extension module is the module where the builder class is defined.

I have also fixed handling of built-in builders and made sure that specifying the extension in ``conf.py/extensions`` still works in case the builder does provide the entry point.

tk0miya commented:
Sorry for very very very late response.
I understand why this only loads builder extensions.
Personally, I love this way. But, at historically, sphinx had avoid to use entry points.
I don't know why, so I can't determine to merge this.

@shimizukawa @birkenfeld could you give us your comments? I think entry_points helps both users and developers certainly. Is there any problem?

tk0miya commented:
Ah, I'd forgotten to comment. the documentation should be updated.
Thanks,

brechtm commented:
Thanks for your feedback @tk0miya. I will await input by @shimizukawa or @birkenfeld before updating the code and documentation.

shimizukawa commented:
Historically(?) reason is here: https://github.com/sphinx-doc/sphinx/issues/1346#issuecomment-70617742

IMO, this feature using entry point doesn't replace a traditional way, so 
it seems good to introduce a new way that using entry point for builder extension.
However, I think it's not necessary to replace the built-in builder registration method with a new method.

tk0miya commented:
> However, I think it's not necessary to replace the built-in builder registration method with a new method.

I can accept that it only support third party builders in Sphinx-1.6 as a first step.
But it should move to new method for all builders including built-in in future version (2.0?).
I believe it brings simple implementation for sphinx.

brechtm commented:
Thanks @shimizukawa, for your input.

I updated the PR. Please review. I have also replied to your review comments.

I wasn't sure where to insert the documentation. This felt like the best place.

Are there any existing tests testing third-party module loading? I'm not sure how to test the entry points, since it requires installation of a distribution.

tk0miya commented on sphinx/application.py:
```diff
@@ -26,6 +26,7 @@
 from docutils import nodes
 from docutils.parsers.rst import convert_directive_function, \
     directives, roles
+from pkg_resources import iter_entry_points
```
Now sphinx does not depend on setuptools. So this might be failed.

@shimizukawa Is it time to add it into dependency list?
Or should we use entry-points only if `pkg_resources` installed?:
```
try:
    import pkg_resources
except ImportError:
    pkg_resources = False

```

brechtm replied:
Right. And `setuptools` needs to be added to `install_requires` in `setup.py` if it is to be a non-optional dependency.

shimizukawa replied:
I agree for depending to setuptools. In fact, most environments have setuptools and troubles derived from setuptools are less than before.
Only one thing, I hear that Ubuntu 16.04 is having trouble with setuptools and pkg_resources.
https://bugs.launchpad.net/ubuntu/+source/python-pip/+bug/1635463
Depending on setuptools, Sphinx may be affected by this bug. However, let's think about after the problem has occurred.

tk0miya commented on doc/extdev/index.rst:
```diff
@@ -22,6 +22,34 @@ The configuration file itself can be treated as an extension if it contains a
+    setup(
+        # ...
+        entry_points={
+            'sphinx.builders': [
```
We have to consider the name of entry_points.
Now Sphinx uses `sphinx_themes` for loading themes.

At same time, we also have to consider the value of entry_points. Is this better to use class name?
But it seems it is not used, its parent module is only used. Is it better to take a module name?

brechtm replied:
I don't know if there are any sort of official guidelines for entry point group names, but the [setuptools](https://setuptools.readthedocs.io/en/latest/setuptools.html#entry-points) / [pkg_resources](https://setuptools.readthedocs.io/en/latest/pkg_resources.html#entry-points) documentation seems to suggest the `sphinx.builders` format. But to remain consistent, you might want to opt for `sphinx_builders`.

The class **is** used; to instantiate the builder which is returned from `create_builder()`. An entry point needs to point to a symbol. Perhaps its possible to point it to a module, but that seems awkward.

tk0miya replied:
Oh, sorry. About the value of entry points, that's my misunderstand.
I thought that `create_builder()` also refers `self.builderclasses` after loading entry points.

shimizukawa replied:
I think that the reason I named `sphinx_themes` was to avoid confusion with the module path. To tell the truth, I was worried too much. So, I do not mind which naming convention.

brechtm replied:
I added a commit that renames the entry point group name to `sphinx_builders`. You may choose whether to merge this commit or not, depending on whether you want `sphinx.builders` or `sphinx_builders`.

tk0miya replied:
+1 for `sphinx.builders`. I also feel the module like name are suggested by the setuptools. As a quick look, `sphinx_themes` is not used so much. So it's a change to rename it.

brechtm replied:
We can support both `sphinx_themes` and `sphinx.themes` for a while and output a deprecation warning when `sphinx_themes` entry points are detected, if you think that would be necessary. But that's for a separate PR...

tk0miya replied:
Don't worry about it. I'll work it later.

tk0miya commented:
ping @shimizukawa

tk0miya commented on sphinx/application.py:
```diff
@@ -298,10 +299,19 @@ def create_builder(self, buildername):
+            else:
+                builderclass = entry_point.load()
+                extension_module = sys.modules[builderclass.__module__]
+                extension_module.setup(self)
```
Just noticed. Please use `sphinx.extensions.load_extension()` to load an extension.

brechtm replied:
Fixed.

brechtm committed:
Use load_extension() to load the builder extension

brechtm committed:
Builder entry points require pkg_resources/setuptools

tk0miya commented on sphinx/application.py:
```diff
@@ -298,10 +299,18 @@ def create_builder(self, buildername):
+                                  ' through entry point' % buildername)
+            else:
+                builderclass = entry_point.load()
+                load_extension(self, builderclass.__module__)
```
Thank you for update, but `load_extension()` expects the module name as a string. So please use `entry_point.module_name` instead.

Inside the function, the builder module is loaded from its name. Hence `entry_point.load()` is not needed here.
How about loading the builder class like following?
```
if buildername not in self.builderclasses:
    entry_points = iter_entry_points('sphinx.builders', buildername)
    try:
        entry_point = next(entry_points)
        load_extension(entry_point.module_name)
    except StopIteration:
        raise ...

builderclass = self.builderclasses[buildername]
return buiderclass(self)
```

brechtm replied:
`builderclass.__module__` is a string as well, so it's equivalent. But I missed the `EntryPoint.module_name`  attribute before and I agree that it is better to use it instead.

I'm was not sure about your suggestion to not make use of `EntryPoint.load()`. It does fit in better with the existing way of handling extensions, but it seems a bit strange to only use the entry point for the module name. But it turns out that an entry point can be just the module:

> [setuptools docs](https://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins): An entry point specifier consists of a name and value, separated by an = sign. The value consists of a dotted module name, **optionally** followed by a : and a dotted identifier naming an object within the module.

`entry_point.load()` then returns the module itself (not the name). So it still doesn't make sense loading the entry point, but I do think this is an improvement since it fits in better with the existing extension handling. Additionally, the builder class now doesn't have to be be defined in the extension module itself (for whatever reason).

PR updated.

brechtm committed:
Do not include the builder class in the entry point

- use the entry point to load the extension module in the usual way
- update the documentation to reflect this change

tk0miya commented:
Merged.

I appreciate your great work! Thank you always!

brechtm commented:
And thank you for the thorough review. I'm very happy with the result!

webknjaz commented on setup.py:
```diff
@@ -52,6 +52,7 @@
     'imagesize',
     'requests>=2.0.0',
     'typing',
+    'setuptools',
```
:man_facepalming: please don't install setuptools from within installed version of setuptools. this causes horrible conflicts.
